### PR TITLE
Fix/refactor

### PR DESCRIPTION
**Pull request by ahmedheakl**
Fix - Teacher, robot envs, and teacher checker refactors

## 📜 Summary
Teacher and robot environment modules have a lot of boilerplate code that needs cleaning. This makes it intimidating to modify or optimize the code. Therefore, in this PR those modules are the main focus of refactoring. 

###### Ticket(s): [optimize-difficulty-calc](https://app.clickup.com/t/8669hwz3c)

- 🔄 **Main changes in this PR**
    - Moved configurations loaded to a different modules with `dataclasses` for configurations modeling. 
    - Add `Position` for storing position instead of passing it as list[int]
    - Refactored all points (coords) to be in type `int`. 
    - Optimized `teacher_checker` module by adding one more guard statement and checking for path validity before getting into the search.
    - Moved `reward` calculation for teacher class into `teacher_utils` module.
    - Changed env width, height from (1080, 720) to (256, 256) for faster compilation and more reliable comparison with previous methods.

- 🐞 **Bug Fixes**:
    - Fixed the problem of the `teacher_checker` function for convex hull difficulty getting stuck forever. 
    - Fixed arguments for `save_log` in `teacher_callback` by swapping its arguments.

- ⚠️ **Breaking changes**
  - `get_region_coords` function doesn't return the lines anymore, it only returns the points defining the region. 
  - All coords are converted from `float` to `int` for faster and more reliable calculations.

## 🚀 How to test this Pull Request

   - Run `pytest test`

## 📝 Notes for future improvements

- Add logger module
- Fix `pd.Dataframe` warning in `teacher_env.opt.episode_statistics`.
- Replace instances of position with a list by the `Position` class.
- Change lists to numpy arrays for faster compilation.
- Fix value of `num_success` as it is viewed in the tensorboard to be fixed at "0".

### CODE REVIEW

File: src/highrl/__init__.py [hunk 1]
```diff
@@ -1,3 +1,5 @@
-# Version of highrl package
-__version__ = "0.1.6"
+"""Setting up pre-call methods and library version"""
+import os
 
+os.environ["TF_CPP_MIN_LOG_LEVEL"] = "3"  # Shutting up tensorflow warnings
+__version__ = "0.1.6"
```
Comment (AbdelrahmanAbdelgwad) [inline src/highrl/__init__.py:5]: update release to 1.0.0

Comment (ahmedheakl) [reply src/highrl/__init__.py]: Done

File: src/highrl/callbacks/teacher_callback.py [hunk 2]
```diff
@@ -114,23 +114,22 @@ def _on_step(self) -> bool:
                 session = DataFrame()
 
                 for env in self.training_env.envs:  # type: ignore
-                    if len(env.session_statistics) != 0:
-                        session = concat(
-                            [session, DataFrame.from_records(env.session_statistics)]
-                        )
+                    if len(env.opt.session_statistics) != 0:
+                        stats_df = DataFrame.from_records(env.opt.session_statistics)
+                        session = concat([session, stats_df])
```
Comment (AbdelrahmanAbdelgwad) [inline src/highrl/callbacks/teacher_callback.py:119]: I think a better naming for the variable stats_df would be session_stats
Also, I think it would be more readable if the variable session was plural ("sessions")

Comment (ahmedheakl) [reply src/highrl/callbacks/teacher_callback.py:119]: Done

File: src/highrl/envs/robot_env.py [hunk 3]
```diff
@@ -314,46 +240,46 @@ def render(
...
             self.image_lock = threading.Lock()
 
-        def make_circle(c, r, res=10):
+        def make_circle(coords: Tuple[int, int], radius: int, res=10) -> np.ndarray:
             """Create circle points
 
             Args:
-                c (list): center of the circle
-                r (float): radius of the circle
+                coords (list): center of the circle
```
Comment (AbdelrahmanAbdelgwad) [inline src/highrl/envs/robot_env.py:272]: rename `coords` to be `center`

File: src/highrl/envs/robot_env.py [hunk 4]
```diff
@@ -314,46 +240,46 @@ def render(
...
+                radius (float): radius of the circle
```
Comment (AbdelrahmanAbdelgwad) [inline src/highrl/envs/robot_env.py:273]: `radius` var should be of type int not float

File: src/highrl/envs/robot_env.py [hunk 5]
```diff
@@ -314,46 +240,46 @@ def render(
...
 
             Returns:
                 list: vertices representing with desired resolution
             """
             thetas = np.linspace(0, 2 * np.pi, res + 1)[:-1]
             verts = np.zeros((res, 2))
-            verts[:, 0] = c[0] + r * np.cos(thetas)
-            verts[:, 1] = c[1] + r * np.sin(thetas)
+            verts[:, 0] = coords[0] + radius * np.cos(thetas)
+            verts[:, 1] = coords[1] + radius * np.sin(thetas)
```
Comment (AbdelrahmanAbdelgwad) [inline src/highrl/envs/robot_env.py:282]: same here with `center` and `coords`

File: src/highrl/envs/robot_env.py [hunk 6]
```diff
@@ -314,46 +240,46 @@ def render(
...
         # Create viewer
         if self.viewer is None:
-            self.viewer = rendering.Viewer(self.width, self.height)
+            self.viewer = rendering.Viewer(self.cfg.width, self.cfg.height)
             self.transform = rendering.Transform()
             self.transform.set_scale(10, 10)
             self.transform.set_translation(128, 128)
             self.score_label = pyglet.text.Label(
                 "0000",
                 font_size=12,
```
Comment (AbdelrahmanAbdelgwad) [inline src/highrl/envs/robot_env.py:249]: Change font size printed so that it is suitable
I think it should be about 5
This should be done for both `score_label` and `iteration_label`

File: src/highrl/envs/robot_env.py (hunk 3 above)
Comment (ahmedheakl) [reply src/highrl/envs/robot_env.py:272]: Done

File: src/highrl/envs/robot_env.py (hunk 4 above)
Comment (ahmedheakl) [reply src/highrl/envs/robot_env.py:273]: Done

File: src/highrl/envs/robot_env.py (hunk 5 above)
Comment (ahmedheakl) [reply src/highrl/envs/robot_env.py:282]: Done

File: src/highrl/envs/robot_env.py (hunk 3 above)
Comment (ahmedheakl) [reply src/highrl/envs/robot_env.py:272]: Done

File: src/highrl/envs/robot_env.py (hunk 4 above)
Comment (ahmedheakl) [reply src/highrl/envs/robot_env.py:273]: Done

File: src/highrl/envs/teacher_env.py [hunk 7]
```diff
@@ -1,233 +1,166 @@
...
+from highrl.utils import training_utils as train_utils
+from highrl.utils import teacher_utils as teach_utils
 
-class TeacherEnv(Env):
 
-    INF_DIFFICULTY = 921600  # w * h = 1280 * 720
+class TeacherEnv(Env):
+    """Environment for training the teacher agent"""
+
+    infinite_difficulty: int = 921600  # w * h = 1280 * 720
```
Comment (AbdelrahmanAbdelgwad) [inline src/highrl/envs/teacher_env.py:19]: Change `infinite_difficulty` to (256 * 256)

File: src/highrl/envs/teacher_env.py [hunk 8]
```diff
@@ -1,233 +1,166 @@
...
+            total_reward += episode_reward
+            total_steps += episode_steps
+            num_success += success_flag
+        self.robot_metrics.success_flag = num_success > 0
+        self.robot_metrics.avg_reward = total_reward / len(self.opt.results)
+        self.robot_metrics.avg_episode_steps = total_steps / len(self.opt.results)
+        self.robot_metrics.success_rate = num_success / len(self.opt.results)
+        print(f"======== Session {self.opt.time_steps} Results ========")
+        results_table = PrettyTable(
+            field_names=["avg_reward", "avg_ep_steps", "success_rate"]
```
Comment (AbdelrahmanAbdelgwad) [inline src/highrl/envs/teacher_env.py:104]: Print `results_table` using `PrettyTable`

File: src/highrl/envs/teacher_env.py [hunk 9]
```diff
@@ -238,408 +171,113 @@ def step(self, action: List) -> Tuple:
...
+        # Flag to advance to next level
+        advance_flag = uniform(0, 1) <= self.cfg.advance_probability
+        self.collect_stats()
 
-        self.robot_level = (self.robot_level + advance_flag) * advance_flag
-        self.robot_env.num_successes = 0
+        self.robot_metrics.level = (
+            self.robot_metrics.level + advance_flag
+        ) * advance_flag
+        self.opt.robot_env.opt.num_successes = 0
```
Comment (AbdelrahmanAbdelgwad) [inline src/highrl/envs/teacher_env.py:249]: I think this line might be why `num_successes` is always zero

File: src/highrl/utils/teacher_utils.py [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+        opt.difficulty_obs = 1
+        is_passed_inf_diff = True
+
+    else:
+        opt.difficulty_area, opt.difficulty_obs = convex_hull_difficulty(
+            opt.robot_env.obstacles,
+            opt.robot_env.robot,
+            opt.width,
+            opt.height,
+        )
```
Comment (AbdelrahmanAbdelgwad) [inline src/highrl/utils/teacher_utils.py:43]: I do not get the logic of the nested if statement

**Review by AbdelrahmanAbdelgwad — APPROVED** [review]
LGTM!! Just reduce the file changed in a single PR next time, so it would be easier for the reviewer to catch bugs and give a clear feedback.

File: src/highrl/envs/teacher_env.py [hunk 11]
```diff
@@ -238,408 +172,113 @@ def step(self, action: List) -> Tuple:
...
-        )
-
-        self.num_successes = int(
-            self.base_num_successes
-            * (self.num_successes_increase_factor) ** self.episodes
-        )
-
-        policy_kwargs = dict(features_extractor_class=Robot1DFeatureExtractor)
-
-        device = self.args.device_used
```
Comment (abdualamer) [inline src/highrl/envs/teacher_env.py:357]: look at the `parser` file

File: src/highrl/utils/parser.py [hunk 12]
```diff
@@ -30,18 +30,18 @@ def parse_args() -> argparse.Namespace:
         action="version",
     )
     parser.add_argument(
-        "--device-used-for-training",
+        "--device",
```
Comment (abdualamer) [inline src/highrl/utils/parser.py:33]: In `teacher_env` it is called `device_used`

File: src/highrl/utils/training_utils.py [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+    if robot_metrics.level == 0:
+        robot_metrics.iid += 1
+
+        print("initiating model ...")
+        model = PPO(
+            "MultiInputPolicy",
+            opt.robot_env,
+            policy_kwargs=policy_kwargs,
+            verbose=2,
+            device=args.device,
```
Comment (abdualamer) [inline src/highrl/utils/training_utils.py:96]: the default device now  is cuda. >> is there any if statement to avoid errors if cuda is not available ?

**Review by abdualamer — COMMENTED** [review]
Really Nice Work
